src/YoutubeAtom.tsx

```diff
@@ -161,273 +45,86 @@ export const YoutubeAtom = ({
-                                const currentTime =
-                                    player.current &&
-                                    (await player.current.getCurrentTime());
-                                const duration =
-                                    player.current &&
-                                    (await player.current.getDuration());
-
-                                if (!duration || !currentTime) return;
-
-                                const percentPlayed =
-                                    (currentTime / duration) * 100;
-
-                                if (!hasSent25Event && 25 < percentPlayed) {
-                                    eventEmitters.forEach((eventEmitter) =>
-                                        eventEmitter('25'),
-                                    );
-                                    hasSent25Event = true;
-                                }
-
-                                if (!hasSent50Event && 50 < percentPlayed) {
-                                    eventEmitters.forEach((eventEmitter) =>
-                                        eventEmitter('50'),
-                                    );
-                                    hasSent50Event = true;
-                                }
-
-                                if (!hasSent75Event && 75 < percentPlayed) {
-                                    eventEmitters.forEach((eventEmitter) =>
-                                        eventEmitter('75'),
-                                    );
-                                    hasSent75Event = true;
-                                }
-
-                                const currentPlayerState =
-                                    player.current &&
-                                    (await player.current.getPlayerState());
-
-                                if (
-                                    currentPlayerState !==
-                                    youtubePlayerState.ENDED
-                                ) {
-                                    // Set a timeout to check progress again in the future
-                                    window.setTimeout(
-                                        () => checkProgress(),
-                                        3000,
-                                    );
-                                }
-                            };
-                        }
-
-                        if (e.data === youtubePlayerState.ENDED) {
-                            eventEmitters.forEach((eventEmitter) =>
-                                eventEmitter('end'),
-                            );
-                        }
-                    },
-                );
-
-            return () => {
-                listener && player.current && player.current.off(listener);
-            };
-        }
-    }, [eventEmitters, loadIframe]);
+    /**
+     * Create a stable callback as it will be a useEffect dependency
```

Comment:
Could you help me understand this in a bit more detail please? What I think I understand so far:
- usecallback ensures we maintain the same function reference between renderings
- there are no dependencies for this usecallback so the memoized function will never be updated
- we pass playerReadyCallback as a prop to the YoutubeAtomPlayer
- onReady is invoked inside the ready listener

^^ what would be the impact of not using usecallback?

Reply:
Sure it is one of those non-obvious, rough edges with `useEffect`.

Oh and sorry the comment could be more specific - it's actually referring to `YoutubeAtomPlayer`'s `useEffect`. I'll update it.

So in `YoutubeAtom` we have the `playerReady` state.

We want to somehow have `YoutubeAtomPlayer` invoke `setPlayerReady` so `YoutubeAtom` knows when to stop showing the overlay or placeholder.

The issue is that if we passed the callback as `onReady={() => setPlayerReady(true)}` that will create a new function on every re-render of `YoutubePlayer`.

That is fine in most cases (e.g. `setOverlayClicked()`), however because we pass that function as a dependency to the `useEffect` within `YoutubeAtomPlayer` the callback function must be 'stable'. Otherwise within the `useEffect` you might get a 'stale' version of the function.

The impact being when the library `youtube-player` tries to call `onReady()` it has a reference to a function which no longer exists. I actually saw this happen - a quick test is to comment out `onReady` from the dependency array and pass in a regular callback and you should see an error in the console.

These links might explain the issue better:
https://dmitripavlutin.com/dont-overuse-react-usecallback/#2-the-purpose-of-usecallback
https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies